provider/oracle/firewall.go

```diff
@@ -0,0 +1,659 @@
+	rulesFilter := []oci.Filter{
+		oci.Filter{
+			Arg:   "dst_list",
+			Value: name,
+		},
+	}
+	rules, err := f.client.AllSecRules(rulesFilter)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	// gsamfira: the oracle compute API does not allow filtering by action
+	ret := []ociResponse.SecRule{}
+	for _, val := range rules.Result {
+		// gsamfira: We set a default policy of DENY. No use in worrying about
+		// DENY rules (if by any chance someone add one manually for some reason)
+		if val.Action != ociCommon.SecRulePermit {
+			continue
+		}
+		// We only care about rules that have a destination set
+		// to a security list. Those lists get attached to VMs
+		// NOTE: someone decided, when writing the oracle API
+		// that some fields should be bool, some should be string.
+		// never mind they both are boolean values...but hey.
+		// I swear...some people like to watch the world burn
+		if val.Dst_is_ip == "true" {
+			continue
+		}
+		// We only care about rules that have an IP list as source
+		if val.Src_is_ip == "false" {
+			continue
+		}
+		ret = append(ret, val)
+	}
+	return ret, nil
+}
+
+func (f *Firewall) getAllIPLists() ([]ociResponse.SecIpList, error) {
+	//user defined IP lists
+	secIpLists, err := f.client.AllSecIpLists(nil)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defaultSecIpLists, err := f.client.AllDefaultSecIpLists(nil)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	allIpLists := []ociResponse.SecIpList{}
+	for _, val := range secIpLists.Result {
+		allIpLists = append(allIpLists, val) //[val.Name] = val
+	}
+	for _, val := range defaultSecIpLists.Result {
+		allIpLists = append(allIpLists, val) //[val.Name] = val
+	}
+	return allIpLists, nil
+}
+
+func (f *Firewall) getAllIPListsAsMap() (map[string]ociResponse.SecIpList, error) {
+	allIps, err := f.getAllIPLists()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	allIpLists := map[string]ociResponse.SecIpList{}
+	for _, val := range allIps {
+		allIpLists[val.Name] = val
+	}
+	return allIpLists, nil
+}
+
+func (f *Firewall) isSecIpList(name string) bool {
+	if strings.HasPrefix(name, "seciplist:") {
+		return true
+	}
+	return false
+}
+
+func (f *Firewall) isSecList(name string) bool {
+	if strings.HasPrefix(name, "seclist:") {
+		return true
+	}
+	return false
+}
+
+func (f *Firewall) ensureApplication(portRange network.PortRange, cache *[]ociResponse.SecApplication) (string, error) {
+	for _, val := range *cache {
+		if val.PortProtocolPair() == portRange.String() {
+			return val.Name, nil
+		}
+	}
+	// We need to create a new application
+	// There is always the chance of a race condition
+	// when it comes to creating new resources.
+	// ie: someone may have already created a matching
+	// application between the time we fetched all of them
+	// and the moment we actually got to create one
+	// Worst thing that can happen is that we have a few duplicate
+	// rules, that we cleanup anyway when we destroy the environment
```

Comment:
We do :).

I was referring to the fact that someone outside of juju would create a security application that matches the one we are trying to create. If we create it here, we clean it up later anyway (or at least we should after we do a commit for that :-) ).

The code as it is written now, reuses security applications and security IP lists. However, if at the time juju performs its checks, a matching application is not found, it will try to create a new one. If someone pushes the submit button with a matching rule, after the check is performed by juju, it will create a new sec application. There is no risk of conflict here, just a duplicate rule, that we use and cleanup later.

There is no way to ensure that you won't get a duplicate as long as more then one actor uses a resource (juju and a human operator or another automation tool).